src/utils/p5Utils.js

```diff
@@ -1,50 +1,178 @@
-      /**
-       * Store the window as a private property regardless of whether p5 is present.
-       * Can also set this property by calling method setP5Instance().
-       * @property {Window | p5 | {p5: p5} | undefined} m_p5Instance
-       * @private
-       */
-      this.m_p5Instance = window;
+      source = window;
+    } else if (typeof globalThis !== "undefined") {
+      source = globalThis;
+    }
+
+    if (!source) return;
+
+    if (isP5Constructor(source.p5)) {
+      this.p5Constructor = source.p5;
+      this.registerPreloads();
+    }
+    if (isP5Extensions(source)) {
+      this.p5Extensions = source;
     }
   }
 
   /**
-   * Set p5 instance globally in order to enable p5 features throughout ml5.
-   * Call this function with the p5 instance when using p5 in instance mode.
-   * @param {p5 | {p5: p5}} p5Instance
+   * @public
+   * Set p5 in order to enable p5 features throughout ml5.
+   * This manual setup is only necessary when importing `p5` as a module
+   * rather than loading it on the window.
+   * Can be used in ml5 unit tests to check p5 behavior.
+   *
+   * @example
+   * import p5 from "p5";
+   * import ml5 "ml5";
+   *
+   * ml5.setP5(p5);
+   *
+   * @param {p5} p5
    */
-  setP5Instance(p5Instance) {
-    this.m_p5Instance = p5Instance;
+  setP5(p5) {
+    if (isP5Constructor(p5)) {
+      this.p5Constructor = p5;
+      this.p5Extensions = p5.prototype;
+      this.registerPreloads();
+    } else {
+      console.warn("Invalid p5 object provided to ml5.setP5().");
+    }
   }
 
   /**
+   * @internal
+   * Pass in the ml5 methods which require p5 preload behavior.
+   * Preload functions must return an object with a property `ready` which is a `Promise`.
+   * Preloading will be set up immediately if p5 is available on the window.
+   * Store the references in case p5 is added later.
+   *
+   * @param {*} ml5Library - the `ml5` variable.
+   * @param {Array<string>} methodNames - an array of ml5 functions to preload.
+   */
+  shouldPreload(ml5Library, methodNames) {
+    this.methodsToPreload = methodNames;
+    this.ml5Library = ml5Library;
+    if (this.checkP5()) {
+      this.registerPreloads();
+    }
+  }
+
+  /**
+   * @private
+   * Execute the p5 preload setup using the stored references, provided by shouldPreload().
+   * Won't do anything if `shouldPreload()` has not been called or if p5 is not found.
+   */
+  registerPreloads() {
+    if (this.didSetupPreload) return;
+    const p5 = this.p5Constructor;
+    const ml5 = this.ml5Library;
+    const preloadMethods = this.methodsToPreload;
+    if (!p5 || !ml5) return;
+
+    // Must shallow copy so that it doesn't reference the replaced method.
+    const original = { ...ml5 };
+    // Must alias `this` so that it can be used inside functions with their own `this` context.
+    const self = this;
+
+    // Function to be called when a sketch is created, either in global or instance mode.
+    p5.prototype.myLibInit = function () {
```

Comment:
Just realized that I should rename this to `ml5Init` or something along those lines.  I was calling this `myLibInit` when I was working on the setup in a very generic context for a `myLib` library.

Reply:
@lindapaiste I've tested this and am ready to merge it! Do you want to make this change before I do?

Reply:
Just fixed it!